How many late days are you using for this assignment?

0 days

In [1]:
# Imports and pip installations (if needed)

# In Command Line: pip install pandas

import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
numerical_df = pd.read_csv("data/chronic_kidney_disease_numerical.csv")
categorical_df = pd.read_csv("data/chronic_kidney_disease_categorical.csv")

# Print the data
print("\nNumerical Dataset:")
print(numerical_df)

print("\nCategorical Dataset:")
print(categorical_df)



Numerical Dataset:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

- **Features**: All columns except "Target" are features used to predict kidney disease. These represent various medical measurements and patient characteristics.
- **Label**: The "Target" column is the label, indicating whether the patient has chronic kidney disease ("ckd") or not ("notckd").

The datasets from two csv files contains both categorical and numerical features.

Categorical Features examples:
1. **rbc** (Red Blood Cells): Categorized as "normal" or "abnormal"
2. **htn** (Hypertension): "yes" or "no"
3. **appet** (Appetite): "good" or "poor"

Numerical Features examples:
1. **age**: Patient's age in years
2. **bp**: Blood Pressure
3. **bgr**: Blood Glucose Random

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_num = numerical_df.duplicated(subset=['unique_id'], keep='first').sum()
print(f"Number of duplicate rows in numerical dataset: {duplicate_num}")

# Drop these duplicate rows
if duplicate_num > 0:
    numerical_df = numerical_df.drop_duplicates(subset=['unique_id'], keep='first')
    print(f"After removing duplicates, numerical dataset has {len(numerical_df)} rows")
else:
    print("No duplicates found in numerical dataset")

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_cat = categorical_df.duplicated(subset=['unique_id'], keep='first').sum()
print(f"Number of duplicate rows in categorical dataset: {duplicate_cat}")

if duplicate_cat > 0:
    categorical_df = categorical_df.drop_duplicates(subset=['unique_id'], keep='first')
    print(f"After removing duplicates, categorical dataset has {len(categorical_df)} rows")
else:
    print("No duplicates found in categorical dataset")


Number of duplicate rows in numerical dataset: 30
After removing duplicates, numerical dataset has 390 rows
Number of duplicate rows in categorical dataset: 30
After removing duplicates, categorical dataset has 400 rows


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why <redacted> you think so? If not, what <redacted> the rows that are missing from one of the datasets look like in the combined table?

Answer:

Not all unique IDs are present in both datasets. This is because when examining the data csv files, the categorical dataset has 400 rows while the numerical dataset has 390 row (after removing duplicate rows). This means the categorical dataset contains 10 IDs that don't appear in the numerical dataset.

An outer join preserves all rows from both datasets, filling missing values with NaN when a match isn't found. So, in the combined table, rows missing from numerical dataset would contain categorical data but has NaN values for numerical columns.

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_df = pd.merge(numerical_df, categorical_df, on='unique_id', how='outer')

# Print the combined dataset
print("\nCombined dataset (showing all rows):")
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(combined_df)


Combined dataset (showing all rows):
     unique_id   age     bp    bgr     bu     sc    sod   pot  hemo   pcv  \
0       102677  59.0   70.0   76.0  186.0  15.00  135.0   7.6   7.1  22.0   
1       103570  24.0   80.0  125.0    NaN    NaN  136.0   3.5  15.4  43.0   
2       104675  65.0   80.0  215.0  133.0   2.50    NaN   NaN  13.2  41.0   
3       105923  90.0   90.0  139.0   89.0   3.00  140.0   4.1  12.0  37.0   
4       107297  60.0   70.0    NaN    NaN    NaN    NaN   NaN  16.4  43.0   
5       108785  53.0  100.0  213.0   23.0   1.00  139.0   4.0   NaN   NaN   
6       109053  64.0   60.0  106.0   27.0   0.70  150.0   3.3  14.4  42.0   
7       114717  41.0   80.0  122.0   25.0   0.80  138.0   5.0  17.1  41.0   
8       118191  42.0   70.0   93.0   32.0   0.90  143.0   4.7  16.6  43.0   
9       118596  74.0   60.0   88.0   50.0   0.60  147.0   3.7  17.2  53.0   
10      119638  24.0    NaN  410.0   31.0   1.10    NaN   NaN  12.4  44.0   
11      124304  69.0   70.0  264.0   8

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [8]:
# Calculate the percentage of rows that contain at least one missing value
rows_with_na = combined_df.isna().any(axis=1).sum()
total_rows = len(combined_df)
na_percentage = (rows_with_na / total_rows) * 100

# Print %
print(f"Percentage of rows with at least one missing value: {na_percentage:.2f}%")

# Drop these rows from the dataset
cleaned_df = combined_df.dropna()
print(f"Original dataset shape: {combined_df.shape}")
print(f"Dataset shape after removing rows with missing values: {cleaned_df.shape}")

# Print the entire cleaned dataset
print("\nCleaned dataset:")
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(cleaned_df)

Percentage of rows with at least one missing value: 61.75%
Original dataset shape: (400, 25)
Dataset shape after removing rows with missing values: (153, 25)

Cleaned dataset:
     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  \
0       102677  59.0   70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0   
6       109053  64.0   60.0  106.0   27.0   0.7  150.0   3.3  14.4  42.0   
7       114717  41.0   80.0  122.0   25.0   0.8  138.0   5.0  17.1  41.0   
8       118191  42.0   70.0   93.0   32.0   0.9  143.0   4.7  16.6  43.0   
9       118596  74.0   60.0   88.0   50.0   0.6  147.0   3.7  17.2  53.0   
11      124304  69.0   70.0  264.0   87.0   2.7  130.0   4.0  12.5  37.0   
13      125587  37.0   60.0  109.0   47.0   1.1  141.0   4.9  15.0  48.0   
14      129053  48.0   80.0  133.0  139.0   8.5  132.0   5.5  10.3  36.0   
15      129654  52.0   80.0  125.0   22.0   1.2  139.0   4.6  16.5  43.0   
16      131563  29.0   80.0   83.0   49.0   0.9  139.0   3.3  17

## Part 3.4: Sort the dataset according to the Labels

In [12]:
# Sort the dataset according to the values in 'Target' column
sorted_df = cleaned_df.sort_values(by='Target')

# Reset the indices after sorting
sorted_df = sorted_df.reset_index(drop=True)

# Print the dataset
print("Sorted dataset by Target value:")
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(sorted_df)

Sorted dataset by Target value:
     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  \
0       102677  59.0   70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0   
1       337834  48.0   70.0  117.0   56.0   3.8  111.0   2.5  11.2  32.0   
2       343710  61.0   80.0  173.0  148.0   3.9  135.0   5.2   7.7  24.0   
3       349892  59.0   70.0  424.0   55.0   1.7  138.0   4.5  12.6  37.0   
4       353025  40.0   70.0  253.0  150.0  11.9  132.0   5.6  10.9  31.0   
5       397388  63.0  100.0  280.0   35.0   3.2  143.0   3.5  13.0  40.0   
6       428486  73.0   70.0   70.0   32.0   0.9  125.0   4.0  10.0  29.0   
7       438182  73.0   80.0  253.0  142.0   4.6  138.0   5.8  10.5  33.0   
8       474407  21.0   90.0  107.0   40.0   1.7  125.0   3.5   8.3  23.0   
9       475200  64.0  100.0  163.0   54.0   7.2  140.0   4.6   7.9  26.0   
10      481293  60.0   60.0  288.0   36.0   1.7  130.0   3.0   7.9  25.0   
11      484175  53.0   90.0   70.0  107.0   7.2  114.0  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to <redacted> this **

In [16]:
# Create dummy variables for all categorical columns
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_df = pd.get_dummies(sorted_df, columns=categorical_columns)

# Print the entire dataset
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(encoded_df)

     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  \
0       102677  59.0   70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0   
1       337834  48.0   70.0  117.0   56.0   3.8  111.0   2.5  11.2  32.0   
2       343710  61.0   80.0  173.0  148.0   3.9  135.0   5.2   7.7  24.0   
3       349892  59.0   70.0  424.0   55.0   1.7  138.0   4.5  12.6  37.0   
4       353025  40.0   70.0  253.0  150.0  11.9  132.0   5.6  10.9  31.0   
5       397388  63.0  100.0  280.0   35.0   3.2  143.0   3.5  13.0  40.0   
6       428486  73.0   70.0   70.0   32.0   0.9  125.0   4.0  10.0  29.0   
7       438182  73.0   80.0  253.0  142.0   4.6  138.0   5.8  10.5  33.0   
8       474407  21.0   90.0  107.0   40.0   1.7  125.0   3.5   8.3  23.0   
9       475200  64.0  100.0  163.0   54.0   7.2  140.0   4.6   7.9  26.0   
10      481293  60.0   60.0  288.0   36.0   1.7  130.0   3.0   7.9  25.0   
11      484175  53.0   90.0   70.0  107.0   7.2  114.0   3.7   9.5  29.0   
12      5147

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

Using a single column for binary variables works the same as dummy variables in modeling. However, for categorical variables with more than two values (like five condiment types), a single column would wrongly suggest that there is some sort of order between the values (such as suggesting that condiment type 5 is "more" than type 1, which is wrong). Using dummy variables for multi-class categorical features is better as it creates separate binary columns for each category. The downside is that we'll have more columns in our dataset.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [18]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

df_no_outliers = encoded_df.copy()

for col in numerical_columns:
    mean = df_no_outliers[col].mean()
    std = df_no_outliers[col].std()
    
    lower_bound = mean - 3 * std
    upper_bound = mean + 3 * std
    
    mask = (df_no_outliers[col] >= lower_bound) & (df_no_outliers[col] <= upper_bound)
    
    df_no_outliers = df_no_outliers[mask]

print(f"Original dataset shape: {encoded_df.shape}")
print(f"Dataset shape after removing outliers: {df_no_outliers.shape}")
print(f"Number of rows removed: {encoded_df.shape[0] - df_no_outliers.shape[0]}")

# Print the cleaned dataset without outliers
print("\nCleaned dataset (without outliers):")
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(df_no_outliers)

Original dataset shape: (153, 36)
Dataset shape after removing outliers: (133, 36)
Number of rows removed: 20

Cleaned dataset (without outliers):
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
10      481293  60.0   60.0  288.0   36.0  1.7  130.0  3.0   7.9  25.0   
12      514721  56.0   90.0  129.0  107.0  6.7  131.0  4.8   9.1  29.0   
13      995177  69.0   70.0  214.0   96.0  6.3  120.0  3.9   9.4  28.0   
15      546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0   
18      980291  62.0   70.0  122.0   42.0  1.7  136.0  4.7  12.6  39.0   
19      955830  55.0   80.0  214.0   73

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [ ]:
# Normalize all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

normalized_df = df_no_outliers.copy()

for column in numerical_columns:
    min_val = df_no_outliers[column].min()
    max_val = df_no_outliers[column].max()
    normalized_df[column] = (df_no_outliers[column] - min_val) / (max_val - min_val)

# Print the normalized dataset
print("\nNormalized dataset:")
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(normalized_df)


Normalized dataset:
     unique_id       age    bp       bgr        bu        sc       sod  \
2       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
5       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
7       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
8       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
9       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
10      481293  0.729730  0.00  0.935622  0.169935  0.160494  0.333333   
12      514721  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
13      995177  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
15      546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
18      980291  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
19      955830  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
20      923613  0.783784  0.00  0.725322  0.313725  0.481481  0.566667   
22      843362  0

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

Since unique_id is just a record identifier and has no relationship with the target variable or other features, it should be removed from the dataset to prevent it from interfering with the model's predictions.

In [22]:
# Remove unique_id column
normalized_df = normalized_df.drop('unique_id', axis=1)

# Print the dataset without the unique_id column
print("\nDataset without unique_id column:")
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(normalized_df)


Dataset without unique_id column:
          age    bp       bgr        bu        sc       sod       pot  \
2    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
10   0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.034483   
12   0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.655172   
13   0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
15   0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
18   0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.620690   
19   0.662162  0.50  0.618026  0.411765  0.432099  0.566667  0.689655   
20   0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.862069   
22   0.878378  0

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [24]:
# Export to CSV
normalized_df.to_csv('cleaned_kidney_disease_data.csv', index=False)

# Export to JSON
normalized_df.to_json('cleaned_kidney_disease_data.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

|    |      age |    bp |      bgr |       bu |       sc |       sod |       pot |      hemo |       pcv |      wbcc |   ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|----------|-------|----------|----------|----------|-----------|-----------|-----------|-----------|-----------|-------|--------|---------|------------|-------------|-------|--------|--------|---------|------------|----------------|
|  2 | 0.743243 |  0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000  | 0.793103  | 0.000000  | 0.032258  | 0.395161  | ...   | False  | True    | False      | True        | False | True   | False  | True    | True       | False          |
|  5 | 0.770270 |  1.00 | 0.901288 | 0.163399 | 0.345679 | 0.766667  | 0.206897  | 0.524752  | 0.548387  | 0.443548  | ...   | False  | True    | True       | False       | True  | False  | True   | False   | True       | False          |
|  7 | 0.905405 |  0.50 | 0.785408 | 0.862745 | 0.518519 | 0.600000  | 1.000000  | 0.277228  | 0.322581  | 0.233871  | ...   | False  | True    | True       | False       | True  | False  | True   | False   | True       | False          |
|  8 | 0.202703 |  0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667  | 0.206897  | 0.059406  | 0.000000  | 0.653226  | ...   | True   | False   | True       | False       | True  | False  | False  | True    | True       | False          |
|  9 | 0.783784 |  1.00 | 0.399142 | 0.287582 | 0.839506 | 0.666667  | 0.586207  | 0.019802  | 0.096774  | 0.258065  | ...   | True   | False   | True       | False       | True  | True   | True   | False   | True       | False          |
| 10 | 0.729730 |  0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333  | 0.034483  | 0.019802  | 0.064516  | 0.879032  | ...   | True   | False   | False      | True        | True  | False  | False  | True    | True       | False          |
| 12 | 0.675676 |  0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667  | 0.655172  | 0.138614  | 0.193548  | 0.169355  | ...   | True   | False   | True       | False       | True  | False  | True   | False   | True       | False          |
| 13 | 0.851351 |  0.25 | 0.618026 | 0.562092 | 0.728395 | 0.000000  | 0.344828  | 0.168317  | 0.161290  | 0.580645  | ...   | False  | True    | True       | False       | True  | False  | False  | True    | True       | False          |
| 15 | 0.540541 |  0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000  | 0.379310  | 0.207921  | 0.161290  | 0.830645  | ...   | True   | False   | True       | False       | True  | False  | True   | False   | True       | False          |
| 18 | 0.756757 |  0.25 | 0.223176 | 0.209150 | 0.160494 | 0.533333  | 0.620690  | 0.485149  | 0.516129  | 0.290323  | ...   | True   | False   | True       | False       | True  | False  | True   | False   | True       | False          |
| 19 | 0.662162 |  0.50 | 0.618026 | 0.411765 | 0.432099 | 0.566667  | 0.689655  | 0.316832  | 0.354839  | 0.250000  | ...   | True   | False   | True       | False       | True  | False  | True   | False   | True       | False          |
| 20 | 0.783784 |  0.00 | 0.725322 | 0.313725 | 0.481481 | 0.566667  | 0.862069  | 0.178218  | 0.193548  | 0.258065  | ...   | True   | False   | False      | True        | True  | False  | True   | False   | True       | False          |
| 22 | 0.878378 |  0.00 | 0.206009 | 0.751634 | 0.604938 | 0.533333  | 0.689655  | 0.366337  | 0.387097  | 0.879032  | ...   | True   | False   | False      | True        | True  | False  | True   | False   | True       | False          |
| 23 | 0.878378 |  0.25 | 0.639485 | 0.470588 | 0.395062 | 0.433333  | 0.517241  | 0.267327  | 0.322581  | 0.104839  | ...   | False  | True    | True       | False       | True  | False  | True   | False   | True       | False          |
| 29 | 0.716216 |  0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667  | 0.206897  | 0.267327  | 0.387097  | 0.532258  | ...   | True   | False   | False      | True        | True  | False  | True   | False   | True       | False          |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer:

To mitigate hallucinations during LLM data conversions, we should always validate the output by comparing it with the original data using automated checks (like row counts, column sums, or statistical tests). For larger datasets, it's safer to use LLMs to generate conversion scripts rather than directly manipulate the data, since this allows us to review the logic before applying it to our data.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [30]:
# Create a mapping dictionary from abbreviation to full word
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class',
}

# Create a new DataFrame with renamed columns
renamed_df = normalized_df.rename(columns={k: rename_mapping[k] for k in normalized_df.columns if k in rename_mapping})

# Display the new DataFrame with renamed columns
print(renamed_df)

          age  blood pressure  blood glucose random  blood urea  \
2    0.743243            0.50              0.442060    0.901961   
5    0.770270            1.00              0.901288    0.163399   
7    0.905405            0.50              0.785408    0.862745   
8    0.202703            0.75              0.158798    0.196078   
9    0.783784            1.00              0.399142    0.287582   
..        ...             ...                   ...         ...   
148  0.189189            0.25              0.227468    0.222222   
149  0.216216            0.00              0.115880    0.052288   
150  0.513514            0.00              0.107296    0.235294   
151  0.702703            0.50              0.128755    0.261438   
152  0.689189            0.00              0.150215    0.254902   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
2            0.432099  0.500000   0.793103    0.000000            0.032258   
5            0.345679  0.766667   0.206

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [ ]:
# Converted SQL to Pandas code

# Group by the 'Target' column, count occurrences, and rename the count column
result = normalized_df.groupby('Target_ckd').size().reset_index(name='count')

# Display the result
print(result)

   Target_ckd  count
0       False    111
1        True     22


*Describe what the above code does here*

This code counts how many patients fall into each category (CKD or non-CKD) by grouping the data by the `Target_ckd` column. The output is stored in a new DataFrame called `result` with two columns: `Target_ckd` and `count`.